File: src/challenge05/java/Regex.java
package challenge05.java;

import java.util.regex.Pattern;

/**
 * Clase Regex que contiene expresiones las regulares necesarias para realizar el reto y métodos para validar.
 */
public class Regex 
{
    private final static String ALPHANUMERIC_REGEX = "^[a-zA-Z0-9]+$";
    public final static Pattern ALPHANUMERIC_PATTERN = Pattern.compile(ALPHANUMERIC_REGEX);
    private final static String EMAIL_REGEX = "^(?=.{1,64}@)[A-Za-z0-9_-]+(\\.[A-Za-z0-9_-]+)*@" 
        + "[^-][A-Za-z0-9-]+(\\.[A-Za-z0-9-]+)*(\\.[A-Za-z]{2,})$";
    public final static Pattern EMAIL_PATTERN = Pattern.compile(EMAIL_REGEX);
    private final static String NUMBER_REGEX = "^[0-9]+$";
    public final static Pattern NUMBER_PATTERN = Pattern.compile(NUMBER_REGEX);
    private final static String TEXT_REGEX = "^[a-zA-Z ]+$";
    public final static Pattern TEXT_PATTERN = Pattern.compile(TEXT_REGEX);

    /**
     * Método que verifica si un campo requerido cumple con un patrón dado.
     *
     * @param field   El campo a validar.
     * @param pattern El patrón de expresión regular a comparar.
     * @return true si el campo no está vacío y cumple con el patrón, false de lo contrario.
     */
    public static boolean checkRequiredField(String field, Pattern pattern)
    {
        return !field.isBlank() && pattern.matcher(field).matches();
    }

    /**
     * Método que verifica si un campo opcional cumple con un patrón dado.
     *
     * @param field   El campo a validar.
     * @param pattern El patrón de expresión regular a comparar.
     * @return true si el campo está vacío o cumple con el patrón, false de lo contrario.
     */
    public static boolean checkOptionalField(String field, Pattern pattern)
    {
        return field.isBlank() || pattern.matcher(field).matches();
    }
}

File: src/challenge05/js/Regex.mjs
"use strict";

/**
 * Módulo Regex que contiene las regulares necesarias para realizar el reto y métodos para validar.
 * 
 * Se usan metodos get para que no se puedan modificar las expresiones ya que javascript no tiene soporte
 * para constantes estaticas.
 * @module Regex
 */
export class Regex
{
    static get ALPHANUMERIC_REGEX()
    {
        return /^[a-zA-Z0-9]+$/;
    }

    static get EMAIL_REGEX()
    {
        return /^(?=.{1,64}@)[A-Za-z0-9_-]+(\.[A-Za-z0-9_-]+)*@[A-Za-z0-9-]+(\.[A-Za-z0-9-]+)*(?:\.[A-Za-z]{2,})+$/;
    }

    static get NUMBER_REGEX()
    {
        return /^[0-9]+$/;
    }

    static get TEXT_REGEX()
    {
        return /^[a-zA-Z ]+$/;
    }

    /**
     * Método estático que verifica si un campo requerido cumple con una expresión regular dada.
     *
     * @param {string} field - El campo a validar.
     * @param {RegExp} regex - La expresión regular a comparar.
     * @returns {boolean} - true si el campo no está vacío y cumple con la expresión regular, false de lo contrario.
     * @static
     */
    static checkRequiredField(field, regex)
    {
        if (typeof field !== "string" || !regex instanceof RegExp) 
        {
            return false;            
        }

        return field.length > 0 && regex.test(field);
    }

    /**
     * Método estático que verifica si un campo opcional cumple con una expresión regular dada.
     *
     * @param {string} field - El campo a validar.
     * @param {RegExp} regex - La expresión regular a comparar.
     * @returns {boolean} - true si el campo está vacío o cumple con la expresión regular, false de lo contrario.
     * @static
     */
    static checkOptionalField(field, regex)
    {
        if (typeof field !== "string" || !regex instanceof RegExp) 
        {
            return false;            
        }

        return field.length === 0 || regex.test(field);
    }
}